addons/l10n_sa_edi/models/account_journal.py

```diff
@@ -482,34 +483,49 @@ def _l10n_sa_call_api(self, request_data, request_url, method):
                                                 }, timeout=(30, 30))
             request_response.raise_for_status()
         except (ValueError, HTTPError) as ex:
-            # In the case of an explicit error from ZATCA, i.e we got a response but the code of the response is not 2xx
-            return {
-                'error': _("Server returned an unexpected error: %(error)s", error=(request_response.text or str(ex))),
-                'blocking_level': 'error'
-            }
+            # The 400 case means that it is rejected by ZATCA, but we need to update tha hash as done for accepted.
+            # In the 401+ cases, it is like the server is overloaded e.g. and we still need to resend later.  We do not
+            # erase the index chain (excepted) becauee for ZATCA, one ICV (index chain) needs to correspond to one invoice.
+            if (status_code := ex.response.status_code) != 400:
+                return {
+                    'error': (Markup("<b>[%s]</b>") % status_code) + _("Server returned an unexpected error: %(error)s",
+                               error=(request_response.text or str(ex))),
+                    'blocking_level': 'warning',
+                    'status_code': status_code,
+                    'excepted': True,
+                }
         except RequestException as ex:
             # Usually only happens if a Timeout occurs. In this case we're not sure if the invoice was accepted or
             # rejected, or if it even made it to ZATCA
             return {'error': str(ex), 'blocking_level': 'warning', 'excepted': True}
 
+        if request_response.status_code == '303':
+            return {'error': _('Clearance and reporting seem to have been mixed up. '),
+                    'blocking_level': 'warning', 'excepted': True}
+
         try:
             response_data = request_response.json()
         except json.decoder.JSONDecodeError:
             return {
                 'error': _("JSON response from ZATCA could not be decoded"),
                 'blocking_level': 'error'
             }
+        response_data['status_code'] = request_response.status_code
 
-        if not request_response.ok and (response_data.get('errors') or response_data.get('warnings')):
-            if isinstance(response_data, dict) and response_data.get('errors'):
+        val_res = response_data.get('validationResults', {})
+        if not request_response.ok and (val_res.get('errorMessages') or val_res.get('warningMessages')):
```

Comment:
I will say that I think this method is quite messy. I'm not sure how exactly to improve it.
Like, the `not request_response.ok` implies that we are in a case of a status above or equal 400 => as the above are returned already, the only case left is if the status is 400. 
So we have a part above in the try except and a part here. Of course the answer is json'd before so it can't directly be included above in the try except but you understand my little issue.

Reply:
I agree.  I left it to not to change too much in stable, but this is processed already in multiple places while we somehow just need it in one.  Same for the rejected documents: in the account_edi we can not keep the rejected on the documents, but there are already plans to change l10n_sa_edi to print and send and as we need it as well for PoS, it is good to have documents there as well.